### PR TITLE
Remove GTF_ADDR_ONSTACK and IsVarAddr.

### LINKED ISSUE #22190: [x86][windows] Asserts in `!argx->IsVarAddr() || (argx->gtType == TYP_I_IMPL)`
`JIT_Performance._CodeQuality_Roslyn_CscBench` and  `managed_Compilation` started to fail with `!argx->IsVarAddr() || (argx->gtType == TYP_I_IMPL)`, for example: [jitStress2_jitstressRegs2](https://ci.dot.net/job/dotnet_coreclr/job/master/job/jitstress/job/x86_checked_windows_nt_jitstress2_jitstressregs2/724/testReport/).

There are [29 changes](https://ci.dot.net/job/dotnet_coreclr/job/master/job/jitstress/job/x86_checked_windows_nt_jitstress2_jitstressregs2/724/) in this commit, but I am pretty sure the issue was revealed by #22148.

cc @dotnet/jit-contrib

### PR DESCRIPTION
**Pull request by erozenfeld**
IsVarAddr was checking GTF_ADDR_ONSTACK  to determine if
the GT_ADDR node is an address of a local. This change removes both
GTF_ADDR_ONSTACK and  IsVarAddr and uses IsLocalAdrExpr instead.
IsLocalAddrExpr uses opcodes to determine if GT_ADDR node is
a local address.

GTF_ADDR_ONSTACK flag is ancient, added before 2002 so I couldn't find
the checkin that introduced it.

I changed the assert to a check and an assignment since simplifications
inside fgMorphArgs between
https://github.com/dotnet/coreclr/blob/1a1e4c4d5a8030cb8d82a2e5b06c2ab357b92534/src/jit/morph.cpp#L3709
(which causes https://github.com/dotnet/coreclr/blob/1a1e4c4d5a8030cb8d82a2e5b06c2ab357b92534/src/jit/morph.cpp#L3057)
and
https://github.com/dotnet/coreclr/blob/1a1e4c4d5a8030cb8d82a2e5b06c2ab357b92534/src/jit/morph.cpp#L3790
may result in more GT_ADDR nodes recognized by IsLocalAdrExpr.

x86 and x64 pmi frameworks had no code diffs and some gcinfo reductions
(15 methods with gcinfo diffs in x86).

Fixes #22190.

### CODE REVIEW

**Comment by erozenfeld** [comment]
Typical gcinfo diff:

```diff
-00002C lea      edi, bword ptr [ebp-14CH]
+00002C lea      edi, [ebp-14CH]

-GC info size =  56
+GC info size =  52

 82 34 81 1D 14  | 00C9..00DD  [EBP-134H] a  pointer
-F5 BF 7A        | 0032        reg EDI becoming live (iptr)
+F7 BF 47        | 0047        reg EAX becoming live (iptr)
-3D              | 0037        reg EDI becoming dead
-F1 BF 40        | 0047        reg EAX becoming live (iptr)
 BF 81           | 0048        push ptr  0 (iptr)
```

File: src/jit/importer.cpp [hunk 1]
```diff
@@ -2928,12 +2928,12 @@ CORINFO_CLASS_HANDLE Compiler::impGetObjectClass()
 /* static */
 void Compiler::impBashVarAddrsToI(GenTree* tree1, GenTree* tree2)
 {
-    if (tree1->IsVarAddr())
+    if (tree1->IsLocalAddrExpr() != nullptr)
```
Comment (sandreenko) [inline src/jit/importer.cpp]: Why do you compare `bool IsLocalAddrExpr` with `nullptr`?

Comment (erozenfeld) [reply src/jit/importer.cpp]: It's not `bool`:
https://github.com/dotnet/coreclr/blob/master/src/jit/gentree.h#L1854

File: src/jit/codegenxarch.cpp [hunk 2]
```diff
@@ -4386,9 +4386,11 @@ void CodeGen::genCodeForLclAddr(GenTree* tree)
     regNumber targetReg  = tree->gtRegNum;
 
     // Address of a local var.
-    noway_assert(targetType == TYP_BYREF);
+    noway_assert((targetType == TYP_BYREF) || (targetType == TYP_I_IMPL));
 
-    inst_RV_TT(INS_lea, targetReg, tree, 0, EA_BYREF);
+    emitAttr size = (targetType == TYP_BYREF) ? EA_BYREF : EA_PTRSIZE;
```
Comment (mikedn) [inline src/jit/codegenxarch.cpp]: emitTypeSize(targetType)

File: src/jit/morph.cpp [hunk 3]
```diff
@@ -1619,7 +1619,7 @@ void fgArgInfo::ArgsComplete()
                         case 7:
                             // If we have a stack based LclVar we can perform a wider read of 4 or 8 bytes
                             //
-                            if (argObj->gtObj.gtOp1->IsVarAddr() == false) // Is the source not a LclVar?
+                            if (argObj->gtObj.gtOp1->IsLocalAddrExpr() != nullptr) // Is the source not a LclVar?
```
Comment (mikedn) [inline src/jit/morph.cpp]: == nullptr

Comment (erozenfeld) [reply src/jit/morph.cpp]: Fixed, thanks!

File: src/jit/codegenxarch.cpp (hunk 2 above)
Comment (erozenfeld) [reply src/jit/codegenxarch.cpp]: Changed as suggested. I also added this change to codegenarmarch.

**Comment by erozenfeld** [comment]
This got merged: https://github.com/dotnet/coreclr/commit/4070994640bcc2c4c138b6f695d3fce14ea3efe7
via this PR but somehow github got confused.

**Pull request closed without merging**